exporter/containerimage/attestations.go

```diff
@@ -1,10 +1,184 @@
+			if !errors.Is(err, fs.ErrNotExist) {
+				return err
+			}
+			return nil
+		}
+		f.FileComment = fmt.Sprintf("layerID: %s", desc.Digest.String())
+		return nil
+	}
+	for _, f := range doc.UnpackagedFiles {
+		if err := modifyFile(f); err != nil {
+			return att, err
+		}
+	}
+	for _, p := range doc.Packages {
+		for _, f := range p.Files {
+			if err := modifyFile(f); err != nil {
+				return att, err
+			}
+		}
+	}
+
+	doc.CreationInfo.CreatorTools = append(doc.CreationInfo.CreatorTools, "buildkit-"+version.Version)
+
+	dt, err := encodeSPDX(doc)
+	if err != nil {
+		return att, err
+	}
+
+	return result.Attestation{
+		Kind:        att.Kind,
+		Path:        att.Path,
+		ContentFunc: func() ([]byte, error) { return dt, nil },
+		InToto:      att.InToto,
+	}, nil
+}
+
+func decodeSPDX(dt []byte) (s *spdx.Document2_2, err error) {
+	defer func() {
+		// The spdx tools JSON loader is reported to be panicing sometimes
+		if v := recover(); v != nil {
+			s = nil
+			err = errors.Errorf("an error occurred during SPDX JSON document parsing: %+v", v)
+		}
+	}()
+	doc, err := jsonloader.Load2_2(bytes.NewReader(dt))
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to decode spdx")
+	}
+	return doc, nil
+}
+
+func encodeSPDX(s *spdx.Document2_2) (dt []byte, err error) {
+	defer func() {
+		// The spdx tools JSON saver is reported to be panicing sometimes
+		if v := recover(); v != nil {
+			dt = nil
+			err = errors.Errorf("an error occurred during SPDX JSON document parsing: %+v", v)
+		}
+	}()
+	w := bytes.NewBuffer(nil)
+	err = jsonsaver.Save2_2(s, w)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to encode spdx")
+	}
+	return w.Bytes(), nil
+}
+
+// fileLayerFinder finds the layer that contains a file, with caching to avoid
+// repeated FileList lookups.
+type fileLayerFinder struct {
+	pending []fileLayerEntry
+	cache   map[string]fileLayerEntry
+}
+
+type fileLayerEntry struct {
+	ref  cache.ImmutableRef
+	desc ocispecs.Descriptor
+}
+
+func newFileLayerFinder(target cache.ImmutableRef, remote *solver.Remote) fileLayerFinder {
+	chain := target.LayerChain()
+	descs := remote.Descriptors
+	if len(chain) != len(descs) {
+		panic("ah")
+	}
+
+	pending := []fileLayerEntry{}
+	for i, ref := range chain {
+		pending = append(pending, fileLayerEntry{
+			ref:  ref,
+			desc: descs[i],
+		})
+	}
+	return fileLayerFinder{
+		pending: pending,
+		cache:   map[string]fileLayerEntry{},
+	}
+}
+
+// find finds the layer that contains the file, returning the ImmutableRef and
+// descriptor for the layer.
+func (c *fileLayerFinder) find(ctx context.Context, s session.Group, filename string) (cache.ImmutableRef, *ocispecs.Descriptor, error) {
+	if cache, ok := c.cache[filename]; ok {
+		return cache.ref, &cache.desc, nil
+	}
+	for len(c.pending) > 0 {
+		pending := c.pending[len(c.pending)-1]
+		files, err := pending.ref.FileList(ctx, s)
+		if err != nil {
+			return nil, nil, err
+		}
+		c.pending = c.pending[:len(c.pending)-1]
```

Comment:
I'm not sure I understand how this is supposed to work. It changes the pending array choices for every `find` call.

The point of it should be that `FileList` is only called once per layer access in `fileLayerFinder`. It should save the results in `c`.